### PR TITLE
peer: fix possible deadlock between peer/discMsgStream, gossiper, and…

### PR DESCRIPTION
**Pull request by Roasbeef**
… server

In this commit, we fix a recently discovered deadlock in the peer
struct. At times, the apply() method on the discMsgStream can block due
to the gossiper itself waiting on a response or a mutex. In many cases,
the gossiper is waiting for the server's read mutex as it's attempting
to broadcast a set of messages to new peers. Simultaneously, we may have
a peer that needs to disconnect for various reasons. In this case, it
well attempt to shutdown all active msg streams that are controlled by
the readHandler. However, if the msg stream cannot exit, then the peer
cannot be disconnected, which means the read mutex will never be
released...and deadlock!

We fix this scenario by passing in a quit channel to the discMsgStream.
This channel will allow the readHandler to ensure the disc msg stream
quits properly by passing through its own quit channel.

### CODE REVIEW

File: peer.go [hunk 1]
```diff
@@ -726,9 +724,21 @@ func newDiscMsgStream(p *peer) *msgStream {
 		"Update stream for gossiper created",
 		"Update stream for gossiper exited",
 		1000,
-		func(msg lnwire.Message) {
-			p.server.authGossiper.ProcessRemoteAnnouncement(msg,
-				p.addr.IdentityKey)
+		func(msg lnwire.Message, quitChan chan struct{}) {
+			errChan := p.server.authGossiper.ProcessRemoteAnnouncement(
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply peer.go]: Ahhh, yeah that's true. I'll try to dig of that goroutine dump I obtained...

**Comment by Roasbeef** [comment]
Here's the goroutine dump at play: http://paste.ubuntu.com/p/6Z58tv3rsn/. It was captured by a user when they noticed that certain RPCs were unresponsive. 

If we take a look at it, we'll see several peers attempting to stop their chan msg streams: 
```
4 @ 0x42e9ea 0x42ea9e 0x44ac26 0xc11d6e 0xc12c1d 0x45c2a1
#	0x44ac25	time.Sleep+0x165		/usr/lib/go-1.10/src/runtime/time.go:102
#	0xc11d6d	main.(*msgStream).Stop+0x4d	/home/robban/gocode/src/github.com/lightningnetwork/lnd/peer.go:589
#	0xc12c1c	main.(*peer).readHandler+0x4ec	/home/robban/gocode/src/github.com/lightningnetwork/lnd/peer.go:934
```

When then also see the server attempting to disconnect a peer (due to a concurrent connection), but blocking on the peer's `Disconnect` method: 
```
1 @ 0x42e9ea 0x42ea9e 0x43f2b4 0x43eed9 0x4657c2 0xc117b2 0xc387b1 0xc375cc 0xc581ee 0x45c2a1
#	0x43eed8	sync.runtime_Semacquire+0x38			/usr/lib/go-1.10/src/runtime/sema.go:56
#	0x4657c1	sync.(*WaitGroup).Wait+0x71			/usr/lib/go-1.10/src/sync/waitgroup.go:129
#	0xc117b1	main.(*peer).Disconnect+0x131			/home/robban/gocode/src/github.com/lightningnetwork/lnd/peer.go:477
#	0xc387b0	main.(*server).removePeer+0x130			/home/robban/gocode/src/github.com/lightningnetwork/lnd/server.go:1730
#	0xc375cb	main.(*server).InboundPeerConnected+0x3ab	/home/robban/gocode/src/github.com/lightningnetwork/lnd/server.go:1518
#	0xc581ed	main.(*server).InboundPeerConnected-fm+0x3d	/home/robban/gocode/src/github.com/lightningnetwork/lnd/server.go:498
```

We then also see several peers attempting to exit (in the write handler): 
```
19 @ 0x42e9ea 0x42ea9e 0x43f2b4 0x43eed9 0x4657c2 0xc117b2 0xc16a7c 0x45c2a1
#	0x43eed8	sync.runtime_Semacquire+0x38	/usr/lib/go-1.10/src/runtime/sema.go:56
#	0x4657c1	sync.(*WaitGroup).Wait+0x71	/usr/lib/go-1.10/src/sync/waitgroup.go:129
#	0xc117b1	main.(*peer).Disconnect+0x131	/home/robban/gocode/src/github.com/lightningnetwork/lnd/peer.go:477
#	0xc16a7b	main.(*peer).writeHandler+0x2db	/home/robban/gocode/src/github.com/lightningnetwork/lnd/peer.go:1179
```

Finally, we see the gossiper attempting to broadcast a set of messages, but is blocking as the readlock of the server is already held: 
```
1 @ 0x42e9ea 0x42ea9e 0x43f2b4 0x43eed9 0x465329 0xc34e02 0xc58182 0xbcbb2f 0x45c2a1
#	0x43eed8	sync.runtime_Semacquire+0x38								/usr/lib/go-1.10/src/runtime/sema.go:56
#	0x465328	sync.(*RWMutex).RLock+0x48								/usr/lib/go-1.10/src/sync/rwmutex.go:50
#	0xc34e01	main.(*server).BroadcastMessage+0x41							/home/robban/gocode/src/github.com/lightningnetwork/lnd/server.go:1026
#	0xc58181	main.(*server).BroadcastMessage-fm+0x51							/home/robban/gocode/src/github.com/lightningnetwork/lnd/server.go:356
#	0xbcbb2e	github.com/lightningnetwork/lnd/discovery.(*AuthenticatedGossiper).networkHandler+0xb1e	/home/robban/gocode/src/github.com/lightningnetwork/lnd/discovery/gossiper.go:999
```

**Comment by Roasbeef** [comment]
That final goroutine fragment leads be to believe that we're unable to disconnect a peer within `InboundPeerConnected` (which holds the sever's write lock). As a result, we're unable to continue sending any messages to the disc message stream as that's blocked as well.

**Comment by halseth** [comment]
I think the reason of the deadlock is spot on, however I think the correct solution would be 
1) To pass a `cancel` channel to `apply`, which then would be forwarded to `ProcessRemoteAnnouncement`. 
2) On the send to `d.networkMsgs <- nMsg:` we could select on this `cancel` channel, and return if it is closed.
3) When the `msgStream` is signalled to quit, we would close the `cancel` channel (or just use the `quit` channel as the passed cancel channel for simplicity)

Working from the diff here it would mean passing the channel in `apply` to `ProcessRemoteAnnouncement` and selecting on it there instead.

**Pull request closed without merging**